source/portable/os/ota_os_posix.c

```diff
@@ -150,30 +177,75 @@ OtaOsStatus_t Posix_OtaReceiveEvent( OtaEventContext_t * pEventCtx,
     OtaOsStatus_t otaOsStatus = OtaOsSuccess;
     char * pDst = pEventMsg;
     char buff[ MAX_MSG_SIZE ];
+    struct timespec ts = { 0 };
+    struct pollfd fds = { 0 };
+    int pollResult = 0;
 
     ( void ) pEventCtx;
-    ( void ) timeout;
 
-    /* Receive the next event from OTA event queue.*/
-    errno = 0;
-
-    if( mq_receive( otaEventQueue, buff, sizeof( buff ), NULL ) == -1 )
+    /* Check timeout. */
+    if( timeout > INT32_MAX )
     {
         otaOsStatus = OtaOsEventQueueReceiveFailed;
 
-        LogError( ( "Failed to receive OTA Event: "
-                    "mq_reqeive returned error: "
-                    "OtaOsStatus_t=%i "
-                    ",errno=%s",
-                    otaOsStatus,
-                    strerror( errno ) ) );
+        LogError( ( "Invalid input, Posix_OtaReceiveEvent supports timeout < INT32_MAX, "
+                    "timeout = %u",
+                    timeout ) );
     }
-    else
+
+    /* Check poll result. */
+    if( otaOsStatus == OtaOsSuccess )
     {
-        LogDebug( ( "OTA Event received." ) );
+        /* Receive the next event from OTA event queue.*/
+        errno = 0;
+
+        fds.fd = otaEventQueue;
+        fds.events = POLLIN;
+
+        pollResult = poll( &fds, 1, ( int ) timeout );
+
+        if( pollResult == 0 )
+        {
+            otaOsStatus = OtaOsEventQueueReceiveFailed;
 
-        /* copy the data from local buffer.*/
-        ( void ) memcpy( pDst, buff, MAX_MSG_SIZE );
+            LogDebug( ( "Failed to receive OTA Event before timeout" ) );
+        }
+        else if( ( pollResult < 0 ) || !( fds.revents & POLLIN ) )
+        {
+            otaOsStatus = OtaOsEventQueueReceiveFailed;
+
+            LogError( ( "Failed to receive OTA Event: "
+                        "poll returned error: "
+                        ",errno=%s "
+                        ",revents=%x",
+                        strerror( errno ),
+                        fds.revents ) );
+        }
+        else
+        {
+            /* Do nothing in success case. */
+        }
+    }
+
+    /* Receive the message from queue. */
+    if( otaOsStatus == OtaOsSuccess )
+    {
+        if( mq_timedreceive( otaEventQueue, buff, sizeof( buff ), NULL, &ts ) == -1 )
+        {
+            otaOsStatus = OtaOsEventQueueReceiveFailed;
+
+            LogError( ( "Failed to receive OTA Event: "
```

Comment:
Probably should not Log Error as it can cause noisy logs when OTA Agent is idle.